pallets/nfts/src/features/approvals.rs

```diff
@@ -165,10 +179,197 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
+		CollectionApprovals::<T, I>::try_mutate_exists(
+			(&collection, &origin, &delegate),
+			|maybe_approval| -> DispatchResult {
+				let deposit_required = T::CollectionApprovalDeposit::get();
+				let mut current_deposit = match maybe_approval.take() {
+					Some((_, deposit)) => deposit,
+					None => Zero::zero(),
+				};
+
+				if current_deposit < deposit_required {
+					T::Currency::reserve(&origin, deposit_required - current_deposit)?;
+					current_deposit = deposit_required;
+				}
+				*maybe_approval = Some((deadline, current_deposit));
+				Ok(())
+			},
+		)?;
+
+		Self::deposit_event(Event::TransferApproved {
+			collection,
+			item: None,
+			owner: origin,
+			delegate,
+			deadline,
+		});
+
+		Ok(())
+	}
+
+	/// Cancels the transfer of items in the collection that owned by the origin to a delegate.
+	///
+	/// This function is used to cancel the approval for the transfer of items in the `collection`
+	/// that owned by the `origin` to a `delegate`. After canceling the approval, the function
+	/// returns the `origin` back the deposited fund and emits the `ApprovalCancelled` event.
+	///
+	/// - `origin`: The account grants permission to cancel the transfer.
+	/// - `collection`: The identifier of the collection.
+	/// - `delegate`: The account that was previously allowed to take control of items in the
+	///   collection that owned by the origin.
+	pub(crate) fn do_cancel_collection_approval(
+		origin: T::AccountId,
+		collection: T::CollectionId,
+		delegate: T::AccountId,
+	) -> DispatchResult {
+		let (_, deposit) = CollectionApprovals::<T, I>::take((&collection, &origin, &delegate))
+			.ok_or(Error::<T, I>::UnknownCollection)?;
+
+		T::Currency::unreserve(&origin, deposit);
+
+		Self::deposit_event(Event::ApprovalCancelled {
+			collection,
+			owner: origin,
+			item: None,
+			delegate,
+		});
+
+		Ok(())
+	}
+
+	/// Clears all collection approvals.
+	///
+	/// This function is used to clear all approvals to transfer items in the `collection` that
+	/// owned by the `origin` to a `delegate`. After clearing all approvals, the function returns
+	/// the `origin` back the deposited fund of each collection approval and emits the
+	/// `AllApprovalsCancelled` event.
+	///
+	/// - `origin`: The account grants permission to clear the transfer.
+	/// - `collection`: The collection ID containing the item.
+	/// - `witness_approvals`: Information on the collection approvals cleared. This must be
+	///   correct.
+	pub(crate) fn do_clear_all_collection_approvals(
+		origin: T::AccountId,
+		collection: T::CollectionId,
+		witness_approvals: u32,
+	) -> DispatchResult {
+		let mut removed_approvals: u32 = 0;
+		// Iterate and remove each collection approval, return the deposited fund back to the
+		// `origin`.
+		for (_, (_, deposit)) in CollectionApprovals::<T, I>::drain_prefix((collection, &origin)) {
+			ensure!(removed_approvals < witness_approvals, Error::<T, I>::BadWitness);
+			T::Currency::unreserve(&origin, deposit);
```

Comment:
Can we not add the amount and write to storage at the end of the loop? Not sure whether it will optimise but good to try out.

Reply:
I saw this code is used in the pallet-assets as well: https://github.com/paritytech/polkadot-sdk/blob/445c1c80d438d5b937c347b04e4daa66ad25d878/substrate/frame/assets/src/functions.rs#L895 so I don't think there are any better optimization here. But we could do that if you think that one is better.

Reply:
Something to check for at the end then perhaps, would be interesting to know